frontend/src/packages/editor/operationManager.tsx

```diff
@@ -30,4 +30,98 @@ export const slateToCmsOperation = (editorContent: BlockData, operation: BaseOpe
             noop: {}
         }
     }
+}
+
+
+/** 
+ * The semantics of Slate Operations
+ *      - Slate models operations somewhat weirdly, theres a few key types
+ *          - Insert/Remove text modifies the text field so the "path" isn't actually a complete path in the traditional OT sense 
+ *          - Set-node modifies a specific field
+ *          - Paths like [a, b, c] refer to indexes in either array elements or children fields
+*/
+
+// normalizePath extends the path in a slate operation to also include the field it is editing
+// for example: if we receive the slate operation {set-node underline = true of text-object} with the path [0, 0, 0]
+// the normalised path will be [0, 0, 0, 4] (assuming 4 is index 4 in the text object)
+const normalizePath = (editorContent: BlockData, op: NodeOperation | TextOperation): number[] => {
+    // quirkiness with the CMS text operation interface, the target is included in the path
+    if (op.type === "remove_text" || op.type === "insert_text") {
+        return op.path;
+    }
+
+    // resolve what type we're studying and fetch the field mappings for it
+    return []
+}
+
+const convertTextInsertionOp = (editorContent: BlockData, op: InsertTextOperation): CMSOperation => (
+    {
+        Path: op.path,
+        OperationType: "insert",
+        IsNoOp: false,
+        Operation: {
+            $type: "stringOperation",
+            stringOperation: {
+                rangeStart: op.offset,
+                rangeEnd: op.offset + op.text.length,
+                newValue: op.text
+            }
+        }
+    }
+);
+
+
+const convertTextRemovalOp = (editorContent: BlockData, op: RemoveTextOperation): CMSOperation => (
+    {
+        Path: op.path,
+        OperationType: "delete",
+        IsNoOp: false,
+        Operation: {
+            $type: "stringOperation",
+            stringOperation: {
+                rangeStart: op.offset - op.text.length,
+                rangeEnd: op.offset,
+                newValue: op.text,
+            }
+        }
+    }
+);
+
+
+const normalizeElementPath = (contentBlock: CMSEditorNode, op: NodeOperation | TextOperation): number[] => {
+    switch (true) {
+        case IsCustomElement(contentBlock): return normalizeCustomElementPath(contentBlock as CustomElement, op);
+        case IsCustomTextBlock(contentBlock): return normalizeCustomTextPath(contentBlock as CustomText, op);
+    }
```

Comment:
interesting

i feel like if you want each branch on a single line here it's easier to just do something like
```suggestion
    if (IsCustomElement(contentBlock)) return normalizeCustomElementPath(contentBlock as CustomElement, op);
    if (IsCustomTextBlock(contentBlock)) return normalizeCustomTextPath(contentBlock as CustomText, op);
```

i think this is a go habit?
(note that we don't have to worry about else-if here because they all return anyway)